### ISSUE TITLE
How to projection to different database?

### ISSUE THREAD

**Issue by legendxian** [JasperFx/marten#2451]
How to projection to different database?
Write a database 
Read a other database?

**Comment by gfoidl** [comment]
A strong -- an IMO very good -- concept of Marten is the use of transactional integrity. Thus an external / different database isn't possible by design.

But you can make use of different ways to achieve this:
* outbox pattern (for writing to external stores)
* saga pattern (for coordinated work over different stores)
* ...

**Comment by legendxian** [comment]
> A strong -- an IMO very good -- concept of Marten is the use of transactional integrity. Thus an external / different database isn't possible by design.
> 
> But you can make use of different ways to achieve this:
> 
> * outbox pattern (for writing to external stores)
> * saga pattern (for coordinated work over different stores)
> * ...

But, CQRS said:
one Db to write. 
Another Db to read.
it's High-efficiency
Sorry,My English not good...

**Comment by Hawxy** [comment]
> one Db to write.
> Another Db to read.

This is the type of thing you'd do at the Postgres level by running a replica (which is usually supported out of the box by most managed postgres hosts). You can then use Npgsql's [load balancing support](https://www.npgsql.org/doc/failover-and-load-balancing.html#load-distribution) to split read/writes between primary and replica.

**Comment by gfoidl** [comment]
> But, CQRS said:
> one Db to write.
> Another Db to read.

No, CQRS didn't say this. It's solely about separation of write (commands) and reads (queries). No word about the actual stores -- which BTW can be whatever fits best. Google / Bing for "cqrs myths" and you can read more about this.

If you want to the CQRS-flavor with different write and read dbs, then you could use the outbox-pattern to update the read-db, which contains _materialized views_ / projections for faster querying.

> it's High-efficiency

Maybe, it depends. But for sure it's way more work to develop, maintain, monitor, etc.
So go this route only if needed. To start I'd go with the same db for write and read side, so it's easy to use Marten's projections to create the "materialized views" either inline or async. If well designed a later split to separate write and read dbs isn't that big step then.

---

For general one-write-db and multiple-read-dbs see @Hawxy's answer.

**Comment by jeremydmiller** [comment]
Hey folks, I'm moving this to a discussion just to get it out of my "OMG there's so many Marten issues that need fixing!" view:-)
